src/tests/JIT/HardwareIntrinsics/X86/Shared/GfniAffineTest.template

```diff
@@ -0,0 +1,340 @@
+            {RetBaseType}[] outArray = new {RetBaseType}[RetElementCount];
+
+            Unsafe.CopyBlockUnaligned(ref Unsafe.As<{Op1BaseType}, byte>(ref inArray1[0]), ref Unsafe.AsRef<byte>(left), (uint)Unsafe.SizeOf<{Op1VectorType}<{Op1BaseType}>>());
+            Unsafe.CopyBlockUnaligned(ref Unsafe.As<{Op2BaseType}, byte>(ref inArray2[0]), ref Unsafe.AsRef<byte>(right), (uint)Unsafe.SizeOf<{Op2VectorType}<{Op2BaseType}>>());
+            Unsafe.CopyBlockUnaligned(ref Unsafe.As<{RetBaseType}, byte>(ref outArray[0]), ref Unsafe.AsRef<byte>(result), (uint)Unsafe.SizeOf<{RetVectorType}<{RetBaseType}>>());
+
+            ValidateResult(inArray1, inArray2, outArray, method);
+        }
+
+        private void ValidateResult({Op1BaseType}[] left, {Op2BaseType}[] right, {RetBaseType}[] result, [CallerMemberName] string method = "")
+        {
+            bool succeeded = true;
+
+#pragma warning disable 8321 // emulation methods not used in all test forms
+            static {Op1BaseType} Invert({Op1BaseType} x) => ((ReadOnlySpan<{Op1BaseType}>)[
+                0x00, 0x01, 0x8D, 0xF6, 0xCB, 0x52, 0x7B, 0xD1, 0xE8, 0x4F, 0x29, 0xC0, 0xB0, 0xE1, 0xE5, 0xC7,
+                0x74, 0xB4, 0xAA, 0x4B, 0x99, 0x2B, 0x60, 0x5F, 0x58, 0x3F, 0xFD, 0xCC, 0xFF, 0x40, 0xEE, 0xB2,
+                0x3A, 0x6E, 0x5A, 0xF1, 0x55, 0x4D, 0xA8, 0xC9, 0xC1, 0x0A, 0x98, 0x15, 0x30, 0x44, 0xA2, 0xC2,
+                0x2C, 0x45, 0x92, 0x6C, 0xF3, 0x39, 0x66, 0x42, 0xF2, 0x35, 0x20, 0x6F, 0x77, 0xBB, 0x59, 0x19,
+                0x1D, 0xFE, 0x37, 0x67, 0x2D, 0x31, 0xF5, 0x69, 0xA7, 0x64, 0xAB, 0x13, 0x54, 0x25, 0xE9, 0x09,
+                0xED, 0x5C, 0x05, 0xCA, 0x4C, 0x24, 0x87, 0xBF, 0x18, 0x3E, 0x22, 0xF0, 0x51, 0xEC, 0x61, 0x17,
+                0x16, 0x5E, 0xAF, 0xD3, 0x49, 0xA6, 0x36, 0x43, 0xF4, 0x47, 0x91, 0xDF, 0x33, 0x93, 0x21, 0x3B,
+                0x79, 0xB7, 0x97, 0x85, 0x10, 0xB5, 0xBA, 0x3C, 0xB6, 0x70, 0xD0, 0x06, 0xA1, 0xFA, 0x81, 0x82,
+                0x83, 0x7E, 0x7F, 0x80, 0x96, 0x73, 0xBE, 0x56, 0x9B, 0x9E, 0x95, 0xD9, 0xF7, 0x02, 0xB9, 0xA4,
+                0xDE, 0x6A, 0x32, 0x6D, 0xD8, 0x8A, 0x84, 0x72, 0x2A, 0x14, 0x9F, 0x88, 0xF9, 0xDC, 0x89, 0x9A,
+                0xFB, 0x7C, 0x2E, 0xC3, 0x8F, 0xB8, 0x65, 0x48, 0x26, 0xC8, 0x12, 0x4A, 0xCE, 0xE7, 0xD2, 0x62,
+                0x0C, 0xE0, 0x1F, 0xEF, 0x11, 0x75, 0x78, 0x71, 0xA5, 0x8E, 0x76, 0x3D, 0xBD, 0xBC, 0x86, 0x57,
+                0x0B, 0x28, 0x2F, 0xA3, 0xDA, 0xD4, 0xE4, 0x0F, 0xA9, 0x27, 0x53, 0x04, 0x1B, 0xFC, 0xAC, 0xE6,
+                0x7A, 0x07, 0xAE, 0x63, 0xC5, 0xDB, 0xE2, 0xEA, 0x94, 0x8B, 0xC4, 0xD5, 0x9D, 0xF8, 0x90, 0x6B,
+                0xB1, 0x0D, 0xD6, 0xEB, 0xC6, 0x0E, 0xCF, 0xAD, 0x08, 0x4E, 0xD7, 0xE3, 0x5D, 0x50, 0x1E, 0xB3,
+                0x5B, 0x23, 0x38, 0x34, 0x68, 0x46, 0x03, 0x8C, 0xDD, 0x9C, 0x7D, 0xA0, 0xCD, 0x1A, 0x41, 0x1C
+            ])[x];
+
+            static {RetBaseType} AffineTransform({Op1BaseType} x, ReadOnlySpan<{Op2BaseType}> A, byte b)
```

Comment:
I will point out one last time that the approved managed API shape has argument types that don't match the `gf2p8affine*qb*` and `gf2p8affineinv*qb*` instruction descriptions, and that requires some gymnastics in the test setup as well as in the JIT.  We're effectively testing what would be a signature accepting `VectorXXX<byte>` and `VectorXXX<ReadOnlySpan<byte>>`, where the length of the span is always 8.

This special template is only necessary because of that mismatch between the managed API and what's actually required for testing.

Reply:
I still maintain that the signature is correct.

The definition is:
> For this instruction, an affine transformation is defined by A * x + b where “A” is an 8 by 8 bit matrix, and “x” and “b” are 8-bit vectors

Thus a single vector is represented by `8-bits` and the matrix is represented by 8x 8-bit vectors, thus the most correct signature is
* `public static Vector128<byte> GaloisFieldAffineTransform(Vector128<Vector8<Bit>> x, Vector128<Matrix8x8<Bit>> a, [ConstantExpected] byte b);`

Now we can't define such a signature because the most primitive definable element is a `byte`. We could theoretically instead define `struct Vector8Bit { byte _value; }` and `[InlineArray(8)] struct Matrix8x8Bit { Vector8Bit _e0; }`. However, doing so is extremely complex and goes against the general principles of the how we define and expose these intrinsics. So this ultimately breaks down into the primitive representable element being `byte` and thus us defining `Vector128<byte> x`. We then end up with `Vector128<byte> a` for the matrix for similar reasons.
-- This is no different from `Vector128<float>` being used to represent a `Vector4` and `Vector512<float>` being used to represent a `Matrix4x4`.

There is then a convenience factor/consideration that you may want the same matrix used to transform all vectors and so in many cases you want what is effectively `Vector64<byte> a` where `a` is implicitly "broadcast" up to the full vector width (128/256/512). This is itself complicated for other reasons (there is no actual `Vector64<T>` type on modern x86/x64) so the next closest thing is to take a `Vector128<ulong>` so you can do something like `Vector128.Create(0x012345678ul)`, allowing a single value to be passed in and broadcast up making the code smaller and easier to read.

Exposing an additional overload that takes `Vector128<ulong>` for convenience is potentially reasonable, but it is not "more correct". The special template here is also not required. You could also get implicit broadcast to be tested operate by repeating the bytes in question (more verbose), by utilizing `.AsByte()`, etc. It is also very important that we test the full functionality when the matrix used is not repeated across the different inputs and that is overall more discernable when the "proper" shape is exposed, allowing a matrix to be passed in and specified as (replacing the bits as appropriate):
```csharp
Vector128.Create(
    // first matrix
    0b00000000, // row0
    0b00000000, // row1
    0b00000000, // row2
    0b00000000, // row3
    0b00000000, // row4
    0b00000000, // row5
    0b00000000, // row6
    0b00000000, // row7

    // second matrix
    0b00000000, // row0
    0b00000000, // row1
    0b00000000, // row2
    0b00000000, // row3
    0b00000000, // row4
    0b00000000, // row5
    0b00000000, // row6
    0b00000000, // row7
);
```

Reply:
> We could theoretically instead define struct Vector8Bit { byte _value; } and [InlineArray(8)] struct Matrix8x8Bit { Vector8Bit _e0; }

I don't disagree that something like this would be even more clear, but it's an overcomplication for something that Intel has simply defined as being a quadword.  To summarize:
* The instruction names end in QB, indicating they take qword and byte operands
* The C intrinsic names end in epi64_epi8, indicating they take qword and byte operands
* The pseudocode definitions: `DEFINE affine_byte(tsrc2qw, src1byte, imm8)` indicate they take qword and byte operands.

I have already performed the required gymnastics to make the approved shape work in JIT and in the tests, so I'm fine leaving as is if you are not convinced by seeing them.

> The special template here is also not required. You could also get implicit broadcast to be tested operate by repeating the bytes in question (more verbose)

That's actually what this template does.  Since the other templates use repeated calls to `NextValueOp` to build the test vectors, I set this one up the same, with a method that returns a repeating sequence of 8 bytes to test the constant mask. https://github.com/dotnet/runtime/pull/109537/files/f5854547c5b8cae9df3c1798561c50a461cc011e#diff-80e46f7a2fa37106478ac6f6c63bc113c4e1a7ebb9d6920c2abf54dd43cb6219R111

That's in addition to the tests that use purely random bytes for both arguments.